scanner/matcher/updater/updater.go

```diff
@@ -0,0 +1,366 @@
+		root:           opts.Root,
+		updateInterval: opts.UpdateInterval,
+
+		skipGC:          opts.SkipGC,
+		updateRetention: opts.UpdateRetention,
+	}
+	u.importVulns = func(ctx context.Context, reader io.Reader) error {
+		return libvuln.OfflineImport(ctx, u.pool, reader)
+	}
+
+	u.tryRemoveExiting()
+
+	return u, nil
+}
+
+func fillOpts(opts *Opts) error {
+	if opts == nil {
+		panic("programmer error")
+	}
+	if err := validate(*opts); err != nil {
+		return err
+	}
+
+	if opts.Client == nil {
+		opts.Client = http.DefaultClient
+	}
+	if opts.Root == "" {
+		opts.Root = os.TempDir()
+	}
+	if opts.UpdateRetention <= 1 {
+		opts.UpdateRetention = libvuln.DefaultUpdateRetention
+	}
+	if opts.UpdateInterval < time.Minute {
+		opts.UpdateInterval = 5 * time.Minute
+	}
+
+	return nil
+}
+
+func validate(opts Opts) error {
+	if opts.Store == nil || opts.Locker == nil || opts.Pool == nil || opts.MetadataStore == nil {
+		return errors.New("must provide a Store, a Locker, a Pool, and a MetadataStore")
+	}
+	if _, err := url.Parse(opts.URL); err != nil {
+		return fmt.Errorf("invalid URL: %q", opts.URL)
+	}
+	return nil
+}
+
+// tryRemoveExiting attempts to remove any leftover update files present in the updater's
+// root directory.
+//
+// It is assumed the previous updater used the same root directory.
+// Any errors when attempting to remove pre-existing files are simply logged.
+func (u *Updater) tryRemoveExiting() {
+	ctx := zlog.ContextWithValues(u.ctx, "component", "matcher/updater/Updater.tryRemoveExiting")
+
+	files, err := fs.Glob(os.DirFS(u.root), updateFilePattern)
+	if err != nil {
+		zlog.Warn(ctx).Err(err).Msg("searching for previous update files to remove")
+	}
+	for _, file := range files {
+		if err := os.Remove(filepath.Join(u.root, file)); err != nil {
+			zlog.Warn(ctx).Err(err).Msgf("removing previous update file %s", file)
+		}
+	}
+}
+
+// Close cancels the updater's context and prevents future update cycles.
+// It does **not** cleanup any other resources. It is the user's responsibility to do so.
+func (u *Updater) Close() error {
+	u.cancel()
+	return nil
+}
+
+// Start periodically updates the vulnerability data via Update.
+// Each period is adjusted by some amount of jitter.
+func (u *Updater) Start() error {
+	ctx := zlog.ContextWithValues(u.ctx, "component", "matcher/updater/Updater.Start")
+
+	zlog.Info(ctx).Msg("starting initial update")
+	if err := u.update(ctx); err != nil {
+		zlog.Error(ctx).Err(err).Msg("errors encountered during updater run")
+	}
+	zlog.Info(ctx).Msg("completed initial update")
+
+	timer := time.NewTimer(u.updateInterval + jitter())
+	defer timer.Stop()
+	for {
+		select {
+		case <-timer.C:
+			zlog.Info(ctx).Msg("starting update")
+			if err := u.update(ctx); err != nil {
+				zlog.Error(ctx).Err(err).Msg("errors encountered during updater run")
+			}
+			zlog.Info(ctx).Msg("completed update")
+
+			timer.Reset(u.updateInterval + jitter())
+		case <-ctx.Done():
+			return ctx.Err()
```

Comment:
This doesn't appear reachable - perhaps I'm missing it but not seeing anything close/cancel the context. That by design or is there a timeout / cancel / close expected somewhere?

Reply:
Ahh there it is, excellent.